sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/viewmodel/ScheduleRepository.kt

```diff
@@ -262,112 +210,88 @@ open class ScheduleRepository(context: Context) {
+                }
     }
 
     /**
      * Encapsulate db call functionality into it own function for providing custom mock behavior in tests,
      * and also to remove redundant threading code throughout the class.
      */
-    @VisibleForTesting
-    open fun findSchedule(
-            taskRunUuid: UUID,
-            onNext: Action1<ScheduledActivityEntity>,
-            onError: Action1<Throwable>) {
+    @CheckReturnValue
+    internal fun findSchedule(taskRunUuid: UUID): Single<ScheduledActivityEntity> {
 
         // Null schedule usually means that the developer forgot to call createScheduleTaskRunUuid before launching
         // the task associated with the schedule.
         // Or the user is offline and did a task without having synced the study's schedules.
         // Which, that is okay, the s3 upload will still take place, but we won't be able to update the schedule on bridge.
         // Side effects for this will be, no study reports or finishedOn status for the task will show in the UI.
         // TODO: mdephillips 9/14/2018 message the user there will be no history?
-        val guid = scheduleTaskRunUuidMap[taskRunUuid] ?: run {
-            onError.call(Throwable("No schedule guid found for taskRunUuid $taskRunUuid, " +
+        val guid = scheduleTaskRunUuidMap[taskRunUuid]
+        return if (guid == null) {
+            Single.error(Throwable("No schedule guid found for taskRunUuid $taskRunUuid, " +
                     "are you sure you function createScheduleTaskRunUuid() before running the task?"))
-            return
+        } else {
+            toSingle(scheduleDao.activity(guid).firstOrNull(), Throwable("No schedule found in DB with guid $guid"))
         }
-
-        scheduleRepoSubscriptions.add(Observable.just(scheduleDao)
-                .subscribeOn(Schedulers.io())
-                .subscribe({ dao ->
-                    dao.activity(guid).firstOrNull()?.let {
-                        scheduleOnMain(it, onNext, onError)
-                    } ?: run {
-                        val error = Throwable("No schedule found in DB with guid $guid")
-                        scheduleOnMain(error, onError)
-                    }
-                }, {
-                    scheduleOnMain(it, onError)
-                }))
     }
 
-    /**
-     * Helper function to schedule calls on the main thread
-     */
-    @VisibleForTesting
-    protected open fun scheduleOnMain(
-            schedule: ScheduledActivityEntity,
-            onNext: Action1<ScheduledActivityEntity>,
-            onError: Action1<Throwable>) {
-
-        scheduleRepoSubscriptions.add(Observable.just(schedule)
-                .subscribeOn(AndroidSchedulers.mainThread())
-                .subscribe({
-                    onNext.call(it)
-                }, {
-                    onError.call(it)
-                }))
+    @CheckReturnValue
+    internal fun <T> toSingle(nullableValue: T?, throwableForNull: Throwable): Single<T> {
```

Comment:
Same goes for this, extension function on Single objects

Reply:
Removed